src/PJ_hgridshift.c

```diff
@@ -0,0 +1,145 @@
+    if (0==P)
+        return 0;
+
+    if (0!=P->ctx)
+        pj_ctx_set_errno (P->ctx, errlev);
+
+    return pj_dealloc(P);
+}
+
+
+static void freeup (PJ *P) {
+    freeup_msg (P, 0);
+    return;
+}
+
+
+static XYZ forward_3d(LPZ lpz, PJ *P) {
+    PJ_TRIPLET point;
+    point.lpz = lpz;
+
+    if (P->gridlist != NULL) {
+        /* Only try the gridshift if at least one grid is loaded,
+         * otherwise just pass the coordinate through unchanged. */
+        pj_apply_gridshift_3( P->ctx, P->gridlist,
+                              P->gridlist_count, 1, 1, 0,
+                              &point.xyz.x, &point.xyz.y, &point.xyz.z );
+    }
+
+    return point.xyz;
+}
+
+
+static LPZ reverse_3d(XYZ xyz, PJ *P) {
+    PJ_TRIPLET point;
+    point.xyz = xyz;
+
+    if (P->gridlist != NULL) {
+        /* Only try the gridshift if at least one grid is loaded,
+         * otherwise just pass the coordinate through unchanged. */
+        pj_apply_gridshift_3( P->ctx, P->gridlist,
+                              P->gridlist_count, 0, 1, 0,
+                              &point.xyz.x, &point.xyz.y, &point.xyz.z );
+    }
+
+    return point.lpz;
+}
+
+
+static PJ_OBS forward_obs(PJ_OBS obs, PJ *P) {
+    PJ_OBS point;
+    point.coo.xyz = forward_3d (obs.coo.lpz, P);
+    return point;
+}
+
+
+static PJ_OBS reverse_obs(PJ_OBS obs, PJ *P) {
+    PJ_OBS point;
+    point.coo.lpz = reverse_3d (obs.coo.xyz, P);
+    return point;
+}
+
+
+PJ *PROJECTION(hgridshift) {
+
+    if (!pj_param(P->ctx, P->params, "tgrids").i) {
+        pj_log_error(P, "hgridshift: +grids parameter missing.");
+        return freeup_msg(P, -1);
+    }
+
+    /* Build gridlist. P->gridlist can be empty if +grids only ask for optional grids. */
+    P->gridlist = pj_gridlist_from_nadgrids( P->ctx, pj_param(P->ctx, P->params, "sgrids").s,
+                                             &(P->gridlist_count) );
+
+    /* Was gridlist compiled properly? */
+    if ( pj_ctx_get_errno(P->ctx) ) {
+        pj_log_error(P, "hgridshift: could not find required grid(s).");
+        return freeup_msg(P, -38);
+    }
+
+    P->fwdobs = forward_obs;
+    P->invobs = reverse_obs;
+    P->fwd3d  = forward_3d;
+    P->inv3d  = reverse_3d;
+    P->fwd    = 0;
+    P->inv    = 0;
+
+    P->left  = PJ_IO_UNITS_RADIANS;
+    P->right = PJ_IO_UNITS_RADIANS;
+
+    return P;
+}
+
+
+#ifndef PJ_SELFTEST
+/* selftest stub */
+int pj_hgridshift_selftest (void) {return 0;}
+#else
+int pj_hgridshift_selftest (void) {
+    PJ *P;
+    PJ_OBS expect, a, b;
+    double dist;
+
+    /* fail on purpose: +grids parameter it mandatory*/
+    P = pj_create("+proj=hgridshift");
+    if (0!=P)
+        return 99;
+
+    /* fail on purpose: open non-existing grid */
+    P = pj_create("+proj=hgridshift +grids=nonexistinggrid.gsb");
+    if (0!=P)
+        return 999;
+
+
+    P = pj_create ("+proj=hgridshift +grids=nzgd2kgrid0005.gsb +ellps=GRS80");
+    if (0==P)
+        /* very likely the grid is missing */
```

Comment:
A comment makes more sense at the block level. Consider:
`/* Failure most likely means that the grid is missing */
if (0==P)
    return 10
`